### PR TITLE
Fixes: #14044 - Allow regex renaming of unnamed devices

### LINKED ISSUE #14044: NoneType AttributeError when Bulk Renaming Nameless Devices
### NetBox version

3.6.3

### Python version

3.8

### Steps to Reproduce

1. Create devices without a name
2. Select and Bulk Rename those devices
3. **Disable** Regex
4. Attempt a rename operation

Of note, I am also unable to rename nameless devices with Regex on, but no error is observed.

### Expected Behavior

The renaming of bulk devices operation should be successful.

### Observed Behavior

```python
<class 'AttributeError'>

'NoneType' object has no attribute 'replace'

Python version: 3.8.10
NetBox version: 3.6.3
Plugins: 
  netbox_demo: 0.3.1
```

### PR DESCRIPTION
**Pull request by DanSheps**
### Fixes: #14044 - Allow regex renaming of unnamed devices

* Allow regex renaming of unnamed devices (already allowed actually)
* Catch errors relating to unnamed devices or integrity errors as a result of the rename process

Note: I removed the ^$ requirement as, in my view, as long as there is a valid regex and the regex results in a unique set, we don't need to enforce the ^$ requirement.  For example: `^` or `.*` or `.` would all be equally as valid.

The main changes here catch both IntegrityError's from the DB and a validation error to ensure regex is used.

### CODE REVIEW

File: netbox/dcim/views.py [hunk 1]
```diff
@@ -2122,6 +2122,17 @@ class DeviceBulkRenameView(generic.BulkRenameView):
     filterset = filtersets.DeviceFilterSet
     table = tables.DeviceTable
 
+    def _rename_objects(self, form, selected_objects):
```
Comment (jeremystretch) [inline netbox/dcim/views.py:2125]: Private view methods should not be overridden. Additionally, it's not clear why this logic would apply to devices but not other objects being renamed in bulk.

File: netbox/netbox/views/generic/bulk_views.py [hunk 2]
```diff
@@ -776,6 +776,14 @@ def post(self, request):
...
                             return redirect(self.get_return_url(request))
 
+                except IntegrityError as e:
+                    messages.error(self.request, ", ".join(e.args))
+                    clear_events.send(sender=self)
+
+                except ValidationError as e:
+                    messages.error(self.request, ", ".join(e.messages))
+                    clear_events.send(sender=self)
+
```
Comment (jeremystretch) [inline netbox/netbox/views/generic/bulk_views.py:786]: Are these changes strictly relevant to #14044? If not, we should break them out into a separate bug report.

**Review by jeremystretch — CHANGES_REQUESTED** [review]
Might it be simpler to just replace original values of None with an empty string?

File: netbox/netbox/views/generic/bulk_views.py (hunk 2 above)
Comment (DanSheps) [reply netbox/netbox/views/generic/bulk_views.py:786]: They are.  This will throw a validation error if we don't use regex (We can easily dump that though if we want to go to `None == ''`)

The Integrity error will catch any associated integrity errors.  I tried to fine a way to separate the integrity error out and only capture an integrity error that relates to the name field but it wasn't easily do-able.

**Comment by jeremystretch** [comment]
AFAICT the only change actually needed to address #14044 is to ensure that a null name is always treated as an empty string. (Currently this is true only when regex is enabled.) Have I missed something?

**Comment by DanSheps** [comment]
> AFAICT the only change actually needed to address #14044 is to ensure that a null name is always treated as an empty string. (Currently this is true only when regex is enabled.) Have I missed something?

I can get that fixed up right away then

**Review by jeremystretch — APPROVED** [review]
